### PR TITLE
TraderJoe resolver

### PR DESCRIPTION
**Pull request by iri031**
**Trader Joe resolver.**
Protocol: [https://traderjoexyz.com/lending](https://traderjoexyz.com/lending)
Docs: [https://docs.traderjoexyz.com/main/welcome/master](https://docs.traderjoexyz.com/main/welcome/master)
The Resolver has the following functions:

- `getPosition`: Returns the position of user for the jtokens passed.
- `getPositionAll`: Returns the position of user for all the jtokens of the market.
- [x] Implemented resolver.
- [ ] Deployed resolver.

### CODE REVIEW

File: contracts/protocols/avalanche/traderjoe/helpers.sol [hunk 1]
```diff
@@ -0,0 +1,164 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+import "./interfaces.sol";
+import { DSMath } from "../../../utils/dsmath.sol";
+import "hardhat/console.sol";
```
Comment (pradyuman-verma) [inline contracts/protocols/avalanche/traderjoe/helpers.sol]: When you finally push the code, remove this.

File: contracts/protocols/avalanche/traderjoe/interfaces.sol [hunk 2]
```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: MIT
```
Comment (pradyuman-verma) [inline contracts/protocols/avalanche/traderjoe/interfaces.sol:1]: You forgot to add `pragma solidity ^0.8.6`

**Comment by pradyuman-verma** [comment]
@Richa-iitr  Add functions as `view` if they are not changing state.

**Comment by iri031** [comment]
Screenshot from 2022-03-05 14-09-32

**Comment by iri031** [comment]
> @Richa-iitr Add functions as `view` if they are not changing state.

These functions are calling non-view functions of JoeLens contract, can't change this.

File: contracts/protocols/avalanche/traderjoe/helpers.sol [hunk 3]
```diff
@@ -0,0 +1,137 @@
+//SPDX-License-Identifier: MIT
```
Comment (pradyuman-verma) [inline contracts/protocols/avalanche/traderjoe/helpers.sol:1]: You forgot to add `pragma solidity ^0.8.6`

File: contracts/protocols/avalanche/traderjoe/main.sol [hunk 4]
```diff
@@ -0,0 +1,8 @@
+//SPDX-License-Identifier: MIT
```
Comment (pradyuman-verma) [inline contracts/protocols/avalanche/traderjoe/main.sol:1]: here too

**Comment by iri031** [comment]
_o1
_o2
_o3
_o4
_o5

File: contracts/protocols/avalanche/traderjoe/main.sol [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    function getPosition(address owner, address[] memory jTokens) public returns (UserData memory) {
+        return getTraderjoeData(owner, jTokens);
+    }
+
+    function getPositionAll(address owner) public returns (UserData memory) {
+        JToken[] memory joeTokens = Joetroller(getJoetroller()).getAllMarkets();
+        address[] memory jTokens = new address[](joeTokens.length);
+        for (uint256 i = 0; i < joeTokens.length; i++) {
+            jTokens[i] = address(joeTokens[i]);
+        }
```
Comment (pradyuman-verma) [inline contracts/protocols/avalanche/traderjoe/main.sol]: Let's do this in a separate function in `helpers.sol`

File: contracts/protocols/avalanche/traderjoe/helpers.sol [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                balanceData.underlyingTokenBalance,
+                balanceData.underlyingTokenAllowance,
+                getTokenData(jtoken)
+            );
+        }
+    }
+
+    function getTokenData(JToken jtoken) internal returns (JoeTokenData memory tokenData) {
+        IJoeLens joeLens = IJoeLens(getJoelens());
+        JTokenMetadata memory jData = joeLens.jTokenMetadata(jtoken);
```
Comment (pradyuman-verma) [inline contracts/protocols/avalanche/traderjoe/helpers.sol]:
```suggestion
        JTokenMetadata memory jData =  IJoeLens(getJoelens()).jTokenMetadata(jtoken);
```

File: contracts/protocols/avalanche/traderjoe/helpers.sol [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+    function getTokenPrices(JToken jToken, uint256 decimals)
+        internal
+        view
+        returns (uint256 priceInEth, uint256 priceInUsd)
+    {
+        uint256 price = IPriceOracle(getPriceOracle()).getUnderlyingPrice(jToken);
+        uint256 ethPrice = IPriceOracle(getPriceOracle()).getUnderlyingPrice(JToken(getWethAddress()));
+        // console.log("ethPrice");
+        // console.log(ethPrice);
```
Comment (pradyuman-verma) [inline contracts/protocols/avalanche/traderjoe/helpers.sol]: remove this.

File: contracts/protocols/avalanche/traderjoe/main.sol (hunk 5 above)
Comment (iri031) [reply contracts/protocols/avalanche/traderjoe/main.sol]: Added the function in helpers

File: contracts/protocols/avalanche/traderjoe/helpers.sol (hunk 6 above)
Comment (iri031) [reply contracts/protocols/avalanche/traderjoe/helpers.sol]: Done.

File: contracts/protocols/avalanche/traderjoe/helpers.sol (hunk 7 above)
Comment (iri031) [reply contracts/protocols/avalanche/traderjoe/helpers.sol]: Removed the comments.

**Pull request closed without merging**